optimism/src/keccak/lookups.rs

```diff
@@ -87,93 +89,86 @@ impl<Fp: Field> Lookups for KeccakEnv<Fp> {
         // ROUND LOOKUPS
         {
             // THETA LOOKUPS
-            for i in 0..20 {
-                // Check that ThetaRemainderC < 2^64
-                self.add_lookup(Lookup {
-                    numerator: Signed::read_one(),
-                    table_id: LookupTable::RangeCheck16Lookup,
-                    value: vec![self.keccak_state.theta_remainder_c[i].clone()],
-                });
-                // Check ThetaExpandRotC is the expansion of ThetaDenseRotC
-                self.add_lookup(Lookup {
-                    numerator: Signed::read_one(),
-                    table_id: LookupTable::ResetLookup,
-                    value: vec![
-                        self.keccak_state.theta_dense_rot_c[i].clone(),
-                        self.keccak_state.theta_expand_rot_c[i].clone(),
-                    ],
-                });
-                // Check ThetaShiftC0 is the expansion of ThetaDenseC
-                self.add_lookup(Lookup {
-                    numerator: Signed::read_one(),
-                    table_id: LookupTable::ResetLookup,
-                    value: vec![
-                        self.keccak_state.theta_dense_c[i].clone(),
-                        self.keccak_state.theta_shifts_c[i].clone(),
-                    ],
-                });
-                // Check that the rest of ThetaShiftsC are in the Sparse table
-                for j in 1..SHIFTS {
+            for q in 0..QUARTERS {
+                for x in 0..DIM {
+                    // Check that ThetaRemainderC < 2^64
+                    self.add_lookup(Lookup {
+                        numerator: Signed::read_one(),
+                        table_id: LookupTable::RangeCheck16Lookup,
+                        value: vec![self.remainder_c(x, q).clone()],
+                    });
+                    // Check ThetaExpandRotC is the expansion of ThetaDenseRotC
                     self.add_lookup(Lookup {
                         numerator: Signed::read_one(),
-                        table_id: LookupTable::SparseLookup,
-                        value: vec![self.keccak_state.theta_shifts_c[i + 20 * j].clone()],
+                        table_id: LookupTable::ResetLookup,
+                        value: vec![self.dense_rot_c(x, q), self.expand_rot_c(x, q)],
                     });
+                    // Check ThetaShiftC0 is the expansion of ThetaDenseC
+                    self.add_lookup(Lookup {
+                        numerator: Signed::read_one(),
+                        table_id: LookupTable::ResetLookup,
```

Comment:
This ThetaShiftC0 table id collides with the one just before it, the ThetaExpandRotC, is that OK?

Reply:
There are two ways to design interface for adding and using lookup tables. The current one is that you always add some data into the "bag" by `add_lookup`, and this function kind of acts as `append_lookup`. 

The other semantics that I was thinking of in https://github.com/o1-labs/proof-systems/pull/1652, is that instead of `append_lookup` semantics you actually have `add`: if you `add_lookup` a table with the same id as you did before, it will throw an error. So you need to manually concatenate `value` of the "chunks" before `add_lookup`ing them.